ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -488,12 +499,17 @@ private synchronized boolean changeToFollower(long newTerm, boolean force, Objec
     final boolean metadataUpdated = state.updateCurrentTerm(newTerm);
 
     if (old != RaftPeerRole.FOLLOWER || force) {
-      setRole(RaftPeerRole.FOLLOWER, reason);
+      if (old != RaftPeerRole.LISTENER) {
```

Comment:
It is illegal to have old == LISTENER.  Let's throw an exception in the beginning.  Then we don't have to change other code in this method.
```
@@ -485,6 +498,9 @@ class RaftServerImpl implements RaftServer.Division,
    */
   private synchronized boolean changeToFollower(long newTerm, boolean force, Object reason) {
     final RaftPeerRole old = role.getCurrentRole();
+    if (old == RaftPeerRole.LISTENER) {
+      throw new IllegalStateException("Unexpected role " + old);
+    }
     final boolean metadataUpdated = state.updateCurrentTerm(newTerm);
```